### PR TITLE
Re-add `jit` argument

### LINKED ISSUE #218: Non-jitting OOM issues
Without jitting, some properties as `SinkhornOutput.primal_cost` (using online point cloud geometry), as well as functions `DualPotentials.Transport` or `k_means` cause OOM issues. Previously, jitting was done by default, see #192.
Question is whether to re-introduce the `jit` argument to solvers (would be again True by default), always jit the classes or try another solution.

### PR DESCRIPTION
**Pull request by michalk8**
related #192 
closes #218

### CODE REVIEW

**Comment by michalk8** [comment]
Tests seem to be failing because of different defaults in the functional/class interface (e.g., momentum).
Will remove the functional API first; related #204

**Comment by michalk8** [comment]
Will be handled after #222

**Review by marcocuturi — APPROVED** [review]
LGTM, when issue with functional API is solved.

**Comment by michalk8** [comment]
I removed the assertion on `fused_penalty` because it was being traced, imho it's fine.
However, am not sure about merging this PR, since some tests (e.g., `test_autograd_sinkhorn`, fails when `jit=True` is being passed); they don't fail when jitting the gradient function.

File: tests/solvers/linear/sinkhorn_diff_test.py [hunk 1]
```diff
@@ -277,6 +278,8 @@ def loss_fn(x: jnp.ndarray,
           lse_mode=lse_mode,
           min_iterations=min_iter,
           max_iterations=max_iter,
+          # TODO(cuturi): figure out why implicit diff breaks when `jit=True`
+          jit=False,
```
Comment (michalk8) [inline tests/solvers/linear/sinkhorn_diff_test.py:282]: Here jitting causes the loss/grad to be `NaN`.

File: tests/solvers/linear/sinkhorn_diff_test.py [hunk 2]
```diff
@@ -160,7 +160,8 @@ def test_autograd_sinkhorn(
     def reg_ot(a: jnp.ndarray, b: jnp.ndarray) -> float:
       geom = pointcloud.PointCloud(x, y, epsilon=1e-1)
       prob = linear_problem.LinearProblem(geom, a=a, b=b)
-      solver = sinkhorn.Sinkhorn(lse_mode=lse_mode)
+      # TODO: fails with `jit=True`, investigate
+      solver = sinkhorn.Sinkhorn(lse_mode=lse_mode, jit=False)
```
Comment (michalk8) [inline tests/solvers/linear/sinkhorn_diff_test.py:164]: Jitting here causes the assertion below to fail
```pytb
2023-01-09T15:38:39.8171820Z Mismatched elements: 1 / 1 (100%)
2023-01-09T15:38:39.8172223Z Max absolute difference: 0.04766881
2023-01-09T15:38:39.8172477Z Max relative difference: 1.
2023-01-09T15:38:39.8172702Z  x: array(0., dtype=float32)
2023-01-09T15:38:39.8173087Z  y: array(-0.047669, dtype=float32)
2023-01-09T15:38:40.1837846Z FAILED tests/solvers/quadratic/fgw_t
```
